Add an autotune cache for inductor generated kernels

oulgen opened the pull request:
Summary: Inductor currently has a best config cache for kernels that it generates. This is a local cache done via writing to the file system. This diff takes this local cache to remote by reusing the existing triton caching mechanism built via Memcache internally and Redis externally.

Test Plan:
tested locally using `TORCH_INDUCTOR_AUTOTUNE_REMOTE_CACHE =1`

Look at scuba to verify the local testing: https://fburl.com/scuba/triton_remote_cache/z6pypznk

The plan is to land this diff with this turned off and gradually introduce this.

Differential Revision: D54398076




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

jansel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -829,27 +836,52 @@ def cached_autotune(
+                from triton.runtime.cache import RedisRemoteCacheBackend
+
+                remote_cache = RedisRemoteCacheBackend(key)
+            # we already SHA-1 hash the source contents
```
sha-1 -> sha256

jansel reviewed (changes requested):
Can we add a test for this by mocking out RedisRemoteCacheBackend?

vgrytsun commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -829,27 +836,52 @@ def cached_autotune(
+
+                remote_cache = RedisRemoteCacheBackend(key)
+            # we already SHA-1 hash the source contents
+            remote_cache_key = os.path.basename(filename)
+
+        best_config = load_cached_autotuning(
+            cache_filename, remote_cache, remote_cache_key, configs_hash, configs
+        )
```
nit: you can simplify this a bit if you move remote cache / file operations outside of the `load_cached_autotuning` method
```suggestion
            remote_cache_key = os.path.basename(filename)
            candidate_config = remote_cache.get([remote_cache_key])
       else:
           candidate_config = None
           if os.path.exists(cache_filename):
              with open(cache_filename) as fd:
              candidate_config = json.loads(fd.read())

        best_config = load_cached_autotuning(candidate_config, configs_hash, configs)
```

oulgen replied:
I'd still like to use the cache on disk when it is available, so this would be a behavior change. Having that said this is a function that only gets called from a single place, so unless we plan on calling it from other places, not use how valueable the refactor is (while maintaining the local disk priority)

oulgen commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -829,27 +836,52 @@ def cached_autotune(
+
+                remote_cache = FbMemcacheRemoteCacheBackend(key, is_autotune=True)
+            else:
+                from triton.runtime.cache import RedisRemoteCacheBackend
```
It looks like the PR hasn't merged yet (but it is approved). I pinged the author: https://github.com/openai/triton/pull/2934

oulgen commented:
@jansel added a unit test, `unittest.mock` is pretty neat

jansel commented on test/inductor/test_max_autotune.py:
```diff
@@ -221,6 +222,49 @@ def mm(a, b):
+            torch.compile(mm, dynamic=dynamic)(a, b)
+            reset()
+            torch.compile(mm, dynamic=dynamic)(a, b)
+
```
Add an assert that we had a cache hit?

oulgen replied:
So earlier today I was actually playing around with num_get and num_set in order to do exactly that, but it looks like `reset()` wasn't actually enough, the second/third time we end up skipping code generation, i assume there's some other cache i need to reset or disable. Any chance you know what it is off the top of your head?

jansel replied:
Perhaps it is the on-disk cache?

oulgen replied:
Figured this out, will update the test.

oulgen commented:
@jansel @bertmaher 

While testing thoroughly, I hit one snag with backend hash generation. In some configurations, when we call `driver.active.get_current_target` in `cached_autotuner` (which happens inside a subprocess), nvidia driver throws "Cannot re-initialize CUDA in forked subprocess". Since the backend does not change, I should be able to generate the hash once in main process and re-use it in subprocesses. However, I haven't been able to find a way to do this in Python.

The two options I have are:
1) Write the hash to file in main process, and have children processes read that file
2) Codegen the hash into the inductor_meta of cached_autotune.

Any opinions here?

jansel commented:
@oulgen an environment variable seems best because the on-disk cache might be shared between machines in some setups (or the hardware on a machine could change).

Re-initializing CUDA is something we *really* want to avoid, since it is very slow.  The subprocs should not have a CUDA context.

oulgen commented:
> or the hardware on a machine could change

@jansel perhaps there's something I need to understand better. My current implementation (not uploaded) was to
```
@functools.lru_cache(None)
def triton_backend():
    from triton.compiler.compiler import make_backend
    from triton.runtime.driver import driver

    target = driver.active.get_current_target()
    return make_backend(target)
```
and only call in the main process. Write the value of hash into `inductor_meta['backend_hash']` which would be free from the lru cache.

If I understand you correctly, if machines change, we cannot really cache the backend, and would need to compute it again?
If I were to go with env variable solution, I would need to ensure that it is set in the main process, what would be a good place to run the above to set it then (and later only read the env variable in the subprocs)?

jansel commented:
I was worried about:
1) You write the hardware info to disk
2) Someone swaps out their GPU and reboots (or you have a filesystem shared between machines)
3) Another process reads the cached data from disk and the cached data is wrong.

oulgen committed:
Add an autotune cache for inductor generated kernels (#120963)

Summary:

Inductor currently has a best config cache for kernels that it generates. This is a local cache done via writing to the file system. This diff takes this local cache to remote by reusing the existing triton caching mechanism built via Memcache internally and Redis externally.

Test Plan:
tested locally using `TORCH_INDUCTOR_AUTOTUNE_REMOTE_CACHE=1`

Look at scuba to verify the local testing: https://fburl.com/scuba/triton_remote_cache/z6pypznk

The plan is to land this diff with this turned off and gradually introduce this.

Reviewed By: aakhundov

Differential Revision: D54398076

jeffdaily commented:
This PR broke ROCm CI.  Raised DISABLED issue for affected unit test. 
 https://github.com/pytorch/pytorch/issues/121166

oulgen commented:
Thanks, I'll fix this shortly.

jeffdaily commented:
Had to disable another one.  https://github.com/pytorch/pytorch/issues/121194
